runtimes/dart-2.18/server.dart

```diff
@@ -10,56 +10,121 @@ import 'function_types.dart';
+    if ((req.headers['x-open-runtimes-secret'] ?? '') == '' || (req.headers['x-open-runtimes-secret'] ?? '') != (Platform.environment['OPEN_RUNTIMES_SECRET'] ?? '')) {
+      return shelf.Response(500, body: 'Unauthorized. Provide correct "x-open-runtimes-secret" header.');
     }
-    if (req.headers['x-internal-challenge'] !=
-        Platform.environment['INTERNAL_RUNTIME_KEY']) {
-      return shelf.Response(500, body: 'Unauthorized');
+
+    String bodyString = await req.readAsString();
+    dynamic body = bodyString;
+    String method = req.method;
+    Map<String, dynamic> headers = {};
+
+    for (MapEntry entry in req.headers.entries) {
+      String header = entry.key.toLowerCase();
+      if(!header.startsWith('x-open-runtimes-')) {
+        headers[header] = entry.value;
+      }
+    }
+
+    String contentType = req.headers['content-type'] ?? 'plain/text';
+    if(contentType.contains('application/json')) {
+      if(!bodyString.isEmpty) {
+        body = jsonDecode(bodyString);
+      } else {
+        body = {};
+      }
+    }
+
+    String hostHeader = req.headers['host'] ?? '';
+
+    String path = '/' + req.url.path;
+    String host = '';
+    String scheme = req.headers['x-forwarded-proto'] ?? 'http';
+    String defaultPort = scheme == 'https' ? '443' : '80';
+    int port = int.parse(defaultPort);
+    String queryString = req.url.query;
+    Map<String, String> query = {};
+
+    if(hostHeader.contains(':')) {
+      host = hostHeader.split(':')[0];
+      port = int.parse(hostHeader.split(':')[1]);
+    } else {
+      host = hostHeader;
+      port = int.parse(defaultPort);
+    }
+
+    for(final param in queryString.split("&")) {
+      final parts = param.split("=");
+
+      final key = parts[0];
+      final value = parts.sublist(1).join('=');
+
+      if(key != null && !key.isEmpty) {
+        query[key] = value;
+      }
+    }
+
+    String url = scheme + '://' + host;
+
+    if(port != int.parse(defaultPort)) {
+      url += ':' + port.toString();
     }
-    final headers = {
-      'content-type': 'application/json'
-    };
-    
-    try {
-      final bodystring = await req.readAsString();
-      final body = jsonDecode(bodystring);
-      final request = Request(
-        variables: body['variables'] ?? {},
-        headers: body['headers'] ?? {},
-        payload: body['payload'] ?? '',
-      );
 
-      final response = Response();
-      await runZonedGuarded(
+    url += path;
+
+    if(!queryString.isEmpty) {
+      url += '?' + queryString;
+    }
+
+    Request contextReq = new Request(bodyString: bodyString, body: body, headers: headers, method: method, url: url, path: path, port: port, scheme: scheme, host: host, queryString: queryString, query: query);
+    Response contextRes = new Response();
+    Context context = new Context(contextReq, contextRes);
+
+    String customstd = "";
+
+    dynamic output = null;
+    try {
+      await runZoned(
```

Comment:
Another downside about using this `runZoned()` is the stacktrace ends up including the zone:

```
Exception: My custom error
#0      main (file:///Users/steven/Projects/open-runtimes-open-runtimes/runtimes/dart-2.18/example/lib/main.dart:6:3)
#1      main.<anonymous closure>.<anonymous closure> (file:///Users/steven/Projects/open-runtimes-open-runtimes/runtimes/dart-2.18/server.dart:107:36)
#2      _rootRun (dart:async/zone.dart:1391:13)
#3      _CustomZone.run (dart:async/zone.dart:1293:19)
#4      _runZoned (dart:async/zone.dart:1829:10)
#5      runZoned (dart:async/zone.dart:1751:10)
#6      main.<anonymous closure> (file:///Users/steven/Projects/open-runtimes-open-runtimes/runtimes/dart-2.18/server.dart:105:13)
<asynchronous suspension>
#7      handleRequest (package:shelf/shelf_io.dart:137:16)
<asynchronous suspension>
```

Reply:
Is that a problem? We use runZoned to override print method.

Reply:
https://github.com/open-runtimes/open-runtimes/issues/174
probably this is the problem you mentioned - and indeed runZoned with try catch above should solve this bug